src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/Utilities.cs

```diff
@@ -164,55 +165,15 @@ internal static object GetPrimitiveDefaultValue(Type type)
 
     internal static class ExceptionHelpers
     {
-#if FEATURE_STACK_TRACES
-        private const string prevStackTraces = "PreviousStackTraces";
-#endif
-
         /// <summary>
         /// Updates an exception before it's getting re-thrown so
         /// we can present a reasonable stack trace to the user.
         /// </summary>
-        public static Exception UpdateForRethrow(Exception rethrow)
-        {
-#if FEATURE_STACK_TRACES
-            List<StackTrace> prev;
-
-            // we don't have any dynamic stack trace data, capture the data we can
-            // from the raw exception object.
-            StackTrace st = new StackTrace(rethrow, true);
-
-            if (!TryGetAssociatedStackTraces(rethrow, out prev))
-            {
-                prev = new List<StackTrace>();
-                AssociateStackTraces(rethrow, prev);
-            }
-
-            prev.Add(st);
-
-#endif // FEATURE_STACK_TRACES
-            return rethrow;
-        }
-#if FEATURE_STACK_TRACES
-        /// <summary>
-        /// Returns all the stack traces associates with an exception
-        /// </summary>
-        public static IList<StackTrace> GetExceptionStackTraces(Exception rethrow)
-        {
-            List<StackTrace> result;
-            return TryGetAssociatedStackTraces(rethrow, out result) ? result : null;
-        }
-
-        private static void AssociateStackTraces(Exception e, List<StackTrace> traces)
-        {
-            e.Data[prevStackTraces] = traces;
-        }
-
-        private static bool TryGetAssociatedStackTraces(Exception e, out List<StackTrace> traces)
+        public static Exception UnwrapAndRethrow(TargetInvocationException exception)
         {
-            traces = e.Data[prevStackTraces] as List<StackTrace>;
-            return traces != null;
+            ExceptionDispatchInfo.Capture(exception.InnerException).Throw();
+            return ContractUtils.Unreachable;
```

Comment:
I went back and forth on it. Many of the callers have a pattern like:
```csharp
object ret;
try
{
    // assign ret
}
catch (TargetInvocationException e)
{
    throw UnwrapAndRethrow(e);
}
// use ret
```
Without the use of `throw`, we need to assign a default value to `ret`. Alternatively, we can call a `void`-returning `UnwrapAndRethrow` and follow it by `throw ContractUtils.Unreachable`. The rub is we can't declare this helper method as "always throws, never returns" so data flow analysis knows about this behavior.

Any preference? I can be convinced either way.